dll/win32/browseui/aclistisf.cpp

```diff
@@ -20,83 +21,411 @@
 
 #include "precomp.h"
 
-CACListISF::CACListISF() :
-    m_dwOptions(0)
+CACListISF::CACListISF()
+    : m_dwOptions(ACLO_CURRENTDIR | ACLO_MYCOMPUTER)
+    , m_iNextLocation(LT_DIRECTORY)
+    , m_fShowHidden(FALSE)
 {
 }
 
 CACListISF::~CACListISF()
 {
 }
 
+HRESULT CACListISF::NextLocation()
+{
+    TRACE("NextLocation(%p)\n", this);
+    HRESULT hr;
+    switch (m_iNextLocation)
+    {
+        case LT_DIRECTORY:
+            ++m_iNextLocation;
+            if (!ILIsEmpty(m_pidlCurDir) && (m_dwOptions & ACLO_CURRENTDIR))
+            {
+                hr = SetLocation(m_pidlCurDir);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_DIRECTORY\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_DESKTOP:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_DESKTOP)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DESKTOP, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_DESKTOP\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_MYCOMPUTER:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_MYCOMPUTER)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DRIVES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_MYCOMPUTER\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_FAVORITES:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_FAVORITES)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_FAVORITES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_FAVORITES\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_MAX:
+        default:
+            TRACE("LT_MAX\n");
+            return S_FALSE;
+    }
+}
+
+HRESULT CACListISF::SetLocation(LPCITEMIDLIST pidl)
+{
+    TRACE("SetLocation(%p, %p)\n", this, pidl);
+
+    m_pEnumIDList.Release();
+    m_pShellFolder.Release();
+    m_pidlLocation.Free();
+
+    if (!pidl)
+        return E_FAIL;
+
+    LPITEMIDLIST pidlClone = ILClone(pidl);
```

Comment:
Can use the m_pidlLocation directly here (unless there is a good reason not to?)

Reply:
OK.